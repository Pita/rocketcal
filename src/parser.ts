import * as cheerio from "cheerio";
const chrono = require("chrono-node");

export interface Launch {
  label: string;
  date: Date;
  href?: string;
}

export function parseLaunches(body: string): Launch[] {
  const $ = cheerio.load(body);

  // find upcoming table
  const $upcoming = $(".sidebar_content > table")[0];

  // get table rows, cut away header
  const $upcomingRows = $($upcoming).find("> tbody > tr").slice(1);

  const launches: Array<Launch> = [];

  $upcomingRows.each((_, row) => {
    try {
      const $row = $(row);

      const rowText = $row.children("td").last().text();
      const date = chrono.parseDate(rowText);

      const link = $row.find("td > a");
      const label = link.text();
      const href = link.attr("href");

      launches.push({
        label,
        date,
        href,
      });
    } catch (e) {
      console.error(e);
    }
  });

  return launches;
}
