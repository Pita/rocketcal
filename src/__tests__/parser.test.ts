import { parseLaunches } from "../parser";
import * as fs from "fs";
// import { loadHTML } from "../loadHtml";

describe("parser", () => {
  test("parses correctly static html", () => {
    const html = fs.readFileSync(
      `${__dirname}/__mocks__/previews.html`,
      "utf8"
    );
    const launches = parseLaunches(html);
    expect(launches).toMatchSnapshot();
  });

  // test("parses correctly live html", async () => {
  //   const html = await loadHTML();

  //   const launches = parseLaunches(html);
  //   expect(launches.length).toBeGreaterThan(0);
  // });
});
