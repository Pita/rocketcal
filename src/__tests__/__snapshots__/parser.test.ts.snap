// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser parses correctly static html 1`] = `
Array [
  Object {
    "date": 2020-05-24T17:00:00.000Z,
    "href": "https://everydayastronaut.com/launcherone-test-flight/",
    "label": "LauncherOne | Test Flight",
  },
  Object {
    "date": 2020-05-27T20:33:31.000Z,
    "href": "https://everydayastronaut.com/dragon-2-dm-2-spacexs-first-crewed-mission/",
    "label": "Falcon 9 Block 5 | Crew Dragon DM-2",
  },
  Object {
    "date": 2020-05-28T10:00:00.000Z,
    "href": "https://everydayastronaut.com/starship-sn-4-150-meter-hop/",
    "label": "Starship SN-4 | 150 meter hop",
  },
  Object {
    "date": 2020-05-19T10:00:00.000Z,
    "href": "https://everydayastronaut.com/starlink-7-falcon-9-block-5-prelaunch-preview/",
    "label": "Falcon 9 | Starlink 7",
  },
]
`;
